SCRIPT_NAME=SpliceAI_Cohort
RESULTS_DIR=/PATH/TO/OUTPUT/DIRECTORY
INPUT_VCF=/PATH/TO/INPUT/VCF/file.vcf.gz
SPLICEAI_PRECOMPUTED_ALL_VCF=/PATH/TO/spliceai_scores.masked.all.hg38.vcf.gz
SPLICEAI_PRECOMPUTED_ALL_BASEPATH=/PATH/TO/spliceai_scores.masked.all.hg38_
REFERENCE_FASTA=/PATH/TO/GATK_bundle/Homo_sapiens_assembly38.fasta
REFERENCE_FASTA_HG38=/PATH/TO/GATK_bundle/Homo_sapiens_assembly38.fasta
REFERENCE_FASTA_HG19=/PATH/TO/UCSC/homo_sapiens/hg19/ucsc.hg19.fasta
CHAIN_HG38_HG19=/PATH/TO/hg38ToHg19.over.chain.gz
CLINVAR_VCF=/PATH/TO/clinvar_hg38.vcf.gz
HGMD_VCF=/PATH/TO/hgmd_pro_hg38.vcf.gz
GENES_BED_BASEPATH=/PATH/TO/hg38_TX_START_TX_END_
BRANCHPOINT_MERCER_BED_HG38=/PATH/TO/supp_gr.182899.114_Supplemental_DataS2_hg38.bed
BRANCHPOINT_LABRANCHOR_BED_HG38=/PATH/TO/lstm.gencode_v19.hg19.top_hg38.bed
LCR_BED_HG38=/PATH/TO/gnomAD/LCRFromHengHg38.bed
REPEATMASKER_BED_HG38=/PATH/TO/UCSC_hg38_RepeatMasker.bed
PAR_BED_HG38=/PATH/TO/PseudoautosomalRegions_hg38.bed
GNOMAD2E_DIR=/PATH/TO/gnomAD/v2.1.1_liftover/vcf/exomes
GNOMAD2G_DIR=/PATH/TO/gnomAD/v2.1.1_liftover/vcf/genomes
GNOMAD3_DIR=/PATH/TO/gnomAD/v3.1.2
GNOMAD_CONSTRAINT=/PATH/TO/gnomAD/v2.1.1/constraint/gnomad.v2.1.1.lof_metrics.by_gene.txt.bgz
GNOMAD_COVERAGE=/PATH/TO/gnomAD/v3.1.2/gnomad.genomes.r3.0.1.coverage.summary.tsv.bgz
MANE_GRCH38=/PATH/TO/MANE/v1.0/MANE.GRCh38.v1.0.summary.txt.gz
GENES_CHD_ANNOTATION=/PATH/TO/GeneAnnotation_CHD.xlsx
POOL_NAME=NONE
